### PR TITLE
Jakarta Faces 3.0

### PR DESCRIPTION
**Pull request by arjantijms**
Signed-off-by: arjantijms <arjan.tijms@gmail.com>

## Specification PR template
When creating a specification project release review, create PRs with the content defined as follows.

Include the following in the PR:
- [X] A directory in the form wombat/x.y where x.y is the release major.minor version, and the directory contains the following.
- [X] Specification PDF in the form of jakarta-wombat-spec-x.y.pdf
- [X] Specification HTML in the form of jakarta-wombat-spec-x.y.html
- [X] A specification page named _index.md following the template at:
      https://github.com/jakartaee/specification-committee/blob/master/spec_page_template.md
- [x] For a Progress Review, that sufficient progress has been made on a Compatible Implementation and TCK, to ensure that the spec is implementable and testable.
  - N/A
- [X] For a [Release Review](https://www.eclipse.org/projects/handbook/#release-review), a summary that a Compatible Implementation is complete, passes the TCK, and that the TCK includes sufficient coverage of the specification. The TCK users guide MUST include the instructions to run the compatible implementations used to validate the release.
Instructions MAY be by reference.
   - [x] Updated release record
   - [x] Generated IP Log
   - [x] Email to PMC
   - [x] Start release review by emailing EMO 
- [X] The URL of the OSSRH staging repository for the api, javadoc:
     https://jakarta.oss.sonatype.org/content/repositories/staging/jakarta/faces/jakarta.faces-api/3.0.0/
- [X] The URL of the staging directory on downloads.eclipse.org for the proposed EFTL TCK binary:
      https://download.eclipse.org/ee4j/jakartaee-tck/jakartaee9-eftl/promoted/jakarta-faces-tck-3.0.0.zip
- [X] The URL of the compatibility certification request issue:
      https://github.com/eclipse-ee4j/faces-api/issues/1542
- [X] Specification JavaDoc in the wombat/x.y/apidocs directory. 
If desired, an optional second PR can be created to contain just the JavaDoc in the `apidocs` directory.

Note: If any item does not apply, check it and mark N/A below it.

### CODE REVIEW

**Comment by kwsutter** [comment]
# Spec Review Checklist

1. Spec PR
  - [ ] PR uses [template](https://github.com/jakartaee/specifications/blob/master/pull_request_template.md)
  - [ ] Directory of form {spec}/x.y
  - [ ] PDF of form jakarta-{spec}-spec-x.y.pdf ("-spec" preferred but not required)
  - [ ] HTML of form jakarta-{spec}-spec-x.y.html ("-spec" preferred but not required)
  - [ ] Index page {spec}/x.y/_index.md following [template](https://github.com/jakartaee/specification-committee/blob/master/spec_page_template.md)
  - [ ] Index page {spec}/_index.md following [template](https://github.com/jakartaee/specification-committee/blob/master/spec_index_template.md)
  - [ ] No other files (e.g., no jakarta_ee_logo_schooner_color_stacked_default.png)
  - [ ] Staging repository link of the form https://jakarta.oss.sonatype.org/content/repositories/staging/jakarta/{spec}/jakarta.{spec}-api/x.y.z/
  - [ ] EFTL TCK link of the form http://download.eclipse.org/.../+.zip
  - [ ] Compatibility certification link of the form https://github.com/eclipse-ee4j/{project}/#{issue}
  - [ ] (Optional) Second PR for just apidocs

2. _index.md
  - [ ] Link to project release plan of the form https://projects.eclipse.org/projects/ee4j.{spec}/releases/x.y/plan
  - [ ] Link to spec pdf
  - [ ] Link to spec html
  - [ ] Link to apidocs
  - [ ] Link to final TCK download zip file of the form https://download.eclipse.org/jakartaee/{spec}/x.y/*{spec}-tck-x.y.z.zip (The folder path is required, the file name pattern is preferred.)
  - [ ] Link to API jar file of the form https://search.maven.org/artifact/jakarta.{spec}/jakarta.{spec}-api/x.y.z/jar
  - [ ] Name of and link to at least one Compatible Implementation

3. javadocs
  - [ ] Footer contains Eclipse copyright and link to license
  - [ ] ESFL license is included, usually as doc-files/speclicense.html
  - [ ] no META-INF directory in PR
  - [ ] javadocs-jar artifact matches apidocs (optional for this release)

4. Spec PDF
  - [ ] Correct spec title
  - [ ] Version number of the form x.y, not x.y.z
  - [ ] Correct Eclipse copyright line
  - [ ] No DRAFT or SNAPSHOT
  - [ ] Correct Logo

5. Spec HTML
  - [ ] Same as PDF

6. TCK zip file
  - [ ] README file (optional for this release)
  - [ ] EFTL license file, preferably named LICENSE.md
  - [ ] User's Guide (or equivalent documentation)
  - [ ] How to test the Compatible Implementation(s) listed in _index.md above with the TCK (may be in UG)

7. TCK User's Guide (or equivalent documentation)
  - [ ] Software requirements listed
  - [ ] Installation and configuration described
  - [ ] How to run tests
  - [ ] Where to file challenges

8. Compatibility certification request
  - [ ] Request follows [template](https://github.com/jakartaee/specification-committee/blob/master/compatibility-certification-request.md)
  - [ ] SHA-256 fingerprint matches staged TCK zip file
  - [ ] Request issue has `certification` label.

9. TCK results summary
  - [ ] Page is hosted by Compatible Implementation project
  - [ ] Includes all information from certification request
  - [ ] Summary includes number of tests passed, failed, errors
  - [ ] SHA-256 fingerprint matches staged TCK zip file on cert request
 
10. Update Jakarta EE API jar
  - [ ] Update the Jakarta EE API jar by submitting a PR to the jakartaee-api project that updates the version number of your API jar file.

**Comment by arjantijms** [comment]
CQ for IP log: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=22623

**Comment by arjantijms** [comment]
TCK results PR: https://github.com/eclipse-ee4j/glassfish/pull/23212

**Comment by arjantijms** [comment]
Release bug: https://bugs.eclipse.org/bugs/show_bug.cgi?id=567418

**Comment by arjantijms** [comment]
@Pandrex247 Any progress with looking at this one?

File: faces/3.0/_index.md
Comment (Pandrex247) [inline faces/3.0/_index.md:10]:
```suggestion
* [Jakarta Server Faces 3.0 Release Record](https://projects.eclipse.org/projects/ee4j.faces/releases/3.0.0/plan)
```
As per checklist, this link should be in the form of `https://projects.eclipse.org/projects/ee4j.{spec}/releases/x.y/plan`

File: faces/3.0/_index.md
Comment (Pandrex247) [inline faces/3.0/_index.md:18]:
```suggestion
* [Jakarta Server Faces 3.0 TCK](https://download.eclipse.org/jakartaee/faces/3.0/jakarta-faces-tck-3.0.0.zip)([sig](https://download.eclipse.org/jakartaee/faces/3.0/jakarta-faces-tck-3.0.0.zip.sig),[sha](https://download.eclipse.org/jakartaee/faces/3.0/jakarta-faces-tck-3.0.0.zip.sha256),[pub](https://raw.githubusercontent.com/jakartaee/specification-committee/master/jakartaee-spec-committee.pub))
```
Missing the version in your URLs, as per checklist `Link to final TCK download zip file of the form https://download.eclipse.org/jakartaee/{spec}/x.y/*{spec}-tck-x.y.z.zip (The folder path is required, the file name pattern is preferred.) `

**Review by Pandrex247 — COMMENTED** [review]
As stated in hangouts, I hijacked @kwsutter 's comment and was the one marking the checkboxes 😜

Finished running through it now, and here are my takeaways:
* As you commented on the Compatibility Cert Request issue, the TCK SHAs don't match the staged one any more, and so needs updated in all the appropriate locations.
* You're missing the PR to the [JakartaEE-API](https://github.com/eclipse-ee4j/jakartaee-api) project.
* The public TCK results summary appears to be a dead link on your Compatibility Cert Request issue
  * The one on your merged TCK results PR was also a dead link actually - the file in the GlassFish repo is an md file whereas the link expected html
* The index.md file has a couple of incorrect links
  * Release record missing the `/plan` bit
  * TCK links missing the `x.y` bits

**Comment by arjantijms** [comment]
Jakarta EE-API PR: https://github.com/eclipse-ee4j/jakartaee-api/pull/85

**Comment by arjantijms** [comment]
> The public TCK results summary appears to be a dead link on your Compatibility Cert Request issue
>   The one on your merged TCK results PR was also a dead link actually - the file in the GlassFish repo is an md file whereas the link expected html

The link should now be correct. If I click on the link in the [CR](https://github.com/eclipse-ee4j/mojarra/issues/4742) I go to the [TCK results](https://eclipse-ee4j.github.io/glassfish/certifications/jakarta-faces/3.0/TCK-Results).

**Comment by arjantijms** [comment]
> The index.md file has a couple of incorrect links
> Release record missing the /plan bit
> TCK links missing the x.y bits

Thanks! Both fixed now.

**Comment by arjantijms** [comment]
Started the re-run of the TCK which should finish in about one hour.

**Comment by arjantijms** [comment]
New run completed with the latest SHA:

```
BUILD SUCCESSFUL
Total time: 136 minutes 31 seconds
SHA256_IMPL=aea9613798591152cacf3ffcc5c18a0d91a6836fab83f9705b0195e5bcc67447
SHA256_TCK=67e7acb555ac2c8955b4157118ef86a5fbb57fb9a65e3d0d75992ec83004204e
Product_download=https://download.eclipse.org/ee4j/glassfish/weekly/glassfish-6.0.0-SNAPSHOT-2020-09-27.zip
IMPL_download=https://jakarta.oss.sonatype.org/content/repositories/staging/org/glassfish/jakarta.faces/3.0.0-RC4/jakarta.faces-3.0.0-RC4.jar
TCK_download=https://download.eclipse.org/ee4j/jakartaee-tck/jakartaee9-eftl/staged-900/jakarta-faces-tck-3.0.0.zip
OS4=3.12.0 NAME="Alpine Linux" ID=alpine VERSION_ID=3.12.0 PRETTY_NAME="Alpine Linux v3.12" HOME_URL="https://alpinelinux.org/" BUG_REPORT_URL="https://bugs.alpinelinux.org/"
JDK_VERSION=java version "1.8.0_202" Java(TM) SE Runtime Environment (build 1.8.0_202-b08) Java HotSpot(TM) 64-Bit Server VM (build 25.202-b08, mixed mode)
```

**Comment by arjantijms** [comment]
TCK Result summary SHA update: https://github.com/eclipse-ee4j/glassfish/pull/23220

**Comment by Pandrex247** [comment]
That all looks good Arjan, though the [public TCK results](https://eclipse-ee4j.github.io/glassfish/certifications/jakarta-faces/3.0/TCK-Results) I believe needs updating to match the new SHA.

Also, not on the checklist but done as part of me just checking things, there's still some references to Java EE and javax in the api docs. I don't know ottomh if these are valid references and it's a bit late here for me to start digging but just raising it for visibility for when others come to review. The javax references in apidocs seem to mostly be in reference to Java SE classes so I _think_ are fine.

Example of Java EE reference, from `faces\3.0\apidocs\com\sun\faces\spi\InjectionProvider.html`

> This interface defines an integration point for Java EE vendors.

On the other hand, the renderkit and vdldoc directories are utterly chock-a-block with javax references, which I'm 90% certain are incorrect (e.g. references to `javax.el` classes which are no longer javax in the EL PR) - I don't think these have been updated properly.

**Comment by arjantijms** [comment]
> That all looks good Arjan, though the public TCK results I believe needs updating to match the new SHA.

That's done via the PR above. It was waiting for approval, but Gaurav has just done that, so I can merge it in. At some point the public website should refresh then.

**Comment by arjantijms** [comment]
> Example of Java EE reference, from faces\3.0\apidocs\com\sun\faces\spi\InjectionProvider.html

`com/sun` is actually a kind of exception. It should be addressed, but since it has been there forever I didn't change it. I'll make another pass through it though. Thanks!

**Comment by arjantijms** [comment]
@Pandrex247 
New deploy ready: https://deploy-preview-278--jakartaee-specifications.netlify.app

**Comment by Pandrex247** [comment]
Given it another once over, and I think it's 98% of the way there now.
Just spotted a couple more _tiny_ little niggles.

The SHA in your public TCK results doesn't match 😫
Specifically, the SHA from the log output.

https://eclipse-ee4j.github.io/glassfish/certifications/jakarta-faces/3.0/TCK-Results.html

> TCK Version, digital SHA-256 fingerprint and download URL:
Jakarta Server Faces 3.0, TCK, SHA-256: 67e7acb555ac2c8955b4157118ef86a5fbb57fb9a65e3d0d75992ec83004204e

```
BUILD SUCCESSFUL
Total time: 136 minutes 27 seconds
SHA256_IMPL=aea9613798591152cacf3ffcc5c18a0d91a6836fab83f9705b0195e5bcc67447
SHA256_TCK=******bba2c7ea38babed1e2d94236c4391e01bf3bfed45934dc6505365126644084b0******
```

And the second niggle is that the [api javadocs-jar](https://jakarta.oss.sonatype.org/content/repositories/staging/jakarta/faces/jakarta.faces-api/3.0.0/jakarta.faces-api-3.0.0-javadoc.jar) hasn't been updated with the changes you've pushed - it still has the references to "Java EE vendors".
Aside from that inconsistency I've looked over your updates to the vdldoc and other api directories and they seem good, 

Almost there!

**Comment by Pandrex247** [comment]
Also of note is that according to the PR checklist you haven't emailed the PMC or started a release review by emailing the EMO yet.

**Comment by arjantijms** [comment]
> And the second niggle is that the api javadocs-jar hasn't been updated with the changes you've pushed - it still has the references to "Java EE vendors".

That's a good one, not sure how I could have misses that, but I did. Working on it right now.

**Comment by arjantijms** [comment]
@Pandrex247 
[API Javadocs](https://jakarta.oss.sonatype.org/content/repositories/staging/jakarta/faces/jakarta.faces-api/3.0.0/) should now be fixed.

**Comment by arjantijms** [comment]
@Pandrex247 
[SHA in public results](https://github.com/eclipse-ee4j/glassfish/pull/23226) should now be fixed (pending PR merge)

**Comment by arjantijms** [comment]
Release review is scheduled here: https://projects.eclipse.org/projects/ee4j.faces/reviews/3.0.0-release-review
Release bug: https://bugs.eclipse.org/bugs/show_bug.cgi?id=567418
Ballot bug: https://bugs.eclipse.org/bugs/show_bug.cgi?id=567419
CQ/IP log bug: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=22688

**Review by Pandrex247 — APPROVED** [review]

**Comment by edburns** [comment]
+1 for continuing Jakarta Faces 3.0.

**Comment by arjantijms** [comment]
@edburns 
> +1 for continuing Jakarta Faces 3.0.

Thanks a lot Ed! We're all building on your many, many years of hard work!

File: faces/3.0/_index.md
Comment (kwsutter) [inline faces/3.0/_index.md:25]: The referenced content via this link has not been updated for Mojarra 3.0.0.  It's still referencing 2.3 and it's dependencies.  Where can this Mojarra 3.0.0-RC4 be accessed?  I found this tag (https://github.com/eclipse-ee4j/mojarra/releases/tag/3.0.0-RC4-RELEASE), but there is no binary associated with it that I can find.  It looks like the CR and TCK Results reference this same page. I won't hold up my vote, but this has to be resolved before this PR can be merged.

**Review by kwsutter — CHANGES_REQUESTED** [review]
The references for the Compatible Implementation need to be cleaned up.

File: faces/3.0/_index.md
Comment (arjantijms) [reply faces/3.0/_index.md:25]: RC4 is primarily available here: 

https://jakarta.oss.sonatype.org/content/repositories/staging/org/glassfish/jakarta.faces/3.0.0-RC4/

But I've just added the binaries to the tag: https://github.com/eclipse-ee4j/mojarra/releases/tag/3.0.0-RC4-RELEASE

I'll update them to the link to the binaries.

**Comment by Pandrex247** [comment]
Ballot complete! 🎉🍰🥂🥳🎉

Actions to now be undertaken:

- [x] Adds this final checklist to the main PR.
- [ ] Adds the approved label to the PRs, and sends out the Ballot Summary per [this template](https://github.com/jakartaee/specification-committee/blob/master/spec_finalization_checklist.md#ballot-summary-email-template) to the public Jakarta EE Specification Committee email list
- [ ] Calculates the staged EFTL TCK signature and promotes it to the committee download area using the https://ci.eclipse.org/jakartaee-spec-committee/job/promote-release/ job. Manually editing the jenkins Build Information will help identify the build (ie. Mail 2.0 or CDI 3.0).
- [ ] Merges the specification (and apidocs) PRs, ensuring the "date:" field in the _index.md file has an appropriate value to allow publishing.
- [ ] Updates the specification page with the ballot results. This is normally done via a separate PR that should be reviewed, approved, and merged.
- [ ] Notifies the EMO of the ballot results by email to emo@eclipse-foundation.org. Just forward the ballot summary note sent earlier to the public Spec Committee email list.
- [ ] Creates an issue in the specification project that includes the following checklist for the specification project team:
  * Promotes api staging release promotes the specification api jars to maven central. An example release job script can be found here https://wiki.eclipse.org/MavenReleaseScript.
  * Go through the merged jakarta.ee specification website page to verify all the links are valid.
  * Approve the compatibility request.
  * The compatible implementation project/vendor MUST send an email to tck@eclipse.org for approval of the compatible implementation for trademark usage.
  * Merge any final release branch as appropriate for the branch management for the project.

**Comment by arjantijms** [comment]
> Ballot complete! 🎉🍰🥂🥳🎉

Now we dance! 🕺💃

**Pull request merged**
